core/src/main/java/org/apache/lucene/analysis/miscellaneous/DuplicateByteSequenceSpotter.java

```diff
@@ -0,0 +1,318 @@
+        // buffer starting from the tail
+        int p = nextFreePos;
+
+        // The first tier of nodes are addressed using individual bytes from the
+        // sequence
+        node = node.add(sequenceBuffer[p], 0);
+        p = nextBufferPos(p);
+        node = node.add(sequenceBuffer[p], 1);
+        p = nextBufferPos(p);
+        node = node.add(sequenceBuffer[p], 2);
+
+        // The final 3 bytes in the sequence are represented in an int
+        // where the 4th byte will contain a hit count.
+
+        // Fill last 3 bytes in search buffer
+        for (int i = 0; i < 3; i++) {
+            p = nextBufferPos(p);
+            lastSequenceSearchBuffer[i] = sequenceBuffer[p];
+        }
+        // The last byte is unused
+        assert (lastSequenceSearchBuffer[3] == 0);
+        short hitCount = node.add(byteSequenceAsInt(lastSequenceSearchBuffer));
+        return (short) (hitCount - 1);
+    }
+
+    // packing an array of 4 bytes to an int, big endian
+    private static int byteSequenceAsInt(byte[] bytes) {
+        return bytes[0] << 24 | (bytes[1] & 0xFF) << 16 | (bytes[2] & 0xFF) << 8 | (bytes[3] & 0xFF);
+    }
+
+    private int nextBufferPos(int p) {
+        p++;
+        if (p >= sequenceBuffer.length) {
+            p = 0;
+        }
+        return p;
+    }
+
+    /**
+     * Base class for nodes in the tree. Subclasses are optimised for use at
+     * different locations in the tree - speed-optimized nodes represent
+     * branches near the root while space-optimized nodes are used for deeper
+     * leaves/branches.
+     */
+    abstract class TreeNode {
+
+        TreeNode(byte key, TreeNode parentNode, int depth) {
+            nodesAllocatedByDepth[depth]++;
+        }
+
+        public abstract TreeNode add(byte b, int depth);
+
+        /**
+         * 
+         * @param byteSequence
+         *            a sequence of bytes encoded as an int
+         * @return the number of times the full sequence has been seen (counting
+         *         up to a maximum of 32767).
+         */
+        public abstract short add(int byteSequence);
+    }
+
+    // Node implementation for use at the root of the tree that sacrifices space
+    // for speed.
+    class RootTreeNode extends TreeNode {
+
+        // A null-or-256 sized array that can be indexed into using a byte for
+        // fast access.
+        // Being near the root of the tree it is expected that this is a
+        // non-sparse array.
+        TreeNode[] children;
+
+        RootTreeNode(byte key, TreeNode parentNode, int depth) {
+            super(key, parentNode, depth);
+            bytesAllocated += ROOT_TREE_NODE_OBJECT_SIZE;
+        }
+
+        public TreeNode add(byte b, int depth) {
+            if (children == null) {
+                children = new TreeNode[256];
+                bytesAllocated += (RamUsageEstimator.NUM_BYTES_OBJECT_REF * 256);
+            }
+            int bIndex = b + 128;
```

Comment:
I *think* `0xFF & b` would amount to the same thing and is a bit easier to read, at least it is for me.